### PR TITLE
feat(Teacher): Implement unit tagging

### LINKED ISSUE #1686: feat(Teacher): Implement unit tagging
Teachers should be able to create tags and apply tags to units in the class schedule and personal unit library.

### PR DESCRIPTION
**Pull request by geoffreykwan**
## Changes
- Added the ability to create, edit, and delete tags
- Added the ability to tag units in the run list and personal unit library
- Added the ability to filter units by tag in the run list and personal unit library

## Test
- Test with https://github.com/WISE-Community/WISE-API/pull/276
- Make sure all the tagging functionality works as expected

Closes #1686

### CODE REVIEW

File: src/app/teacher/select-tags/select-tags.component.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import { Tag } from '../../domain/tag';
+import { AbstractTagsMenuComponent } from '../abstract-tags-menu/abstract-tags-menu.component';
+import { SearchBarComponent } from '../../modules/shared/search-bar/search-bar.component';
+
+@Component({
+  selector: 'select-tags',
+  templateUrl: './select-tags.component.html',
+  styleUrls: ['./select-tags.component.scss'],
+  standalone: true,
+  imports: [CommonModule, MatDividerModule, SearchBarComponent, SharedModule]
```
Comment (hirokiterashima) [inline src/app/teacher/select-tags/select-tags.component.ts]: Should we import SharedModule here? It includes things that we don't use like UnlinkGoogleAccountConfirmComponent and UnlinkGoogleAccountPasswordComponent. Instead, can we import only the classes that we need?

File: src/app/teacher/apply-tags-button/apply-tags-button.component.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  imports: [
+    CommonModule,
+    MatButtonModule,
+    MatDividerModule,
+    MatIconModule,
+    MatMenuModule,
+    MatTooltipModule,
+    SearchBarComponent,
+    SelectAllItemsCheckboxComponent,
+    SharedModule
```
Comment (hirokiterashima) [inline src/app/teacher/apply-tags-button/apply-tags-button.component.ts]: Should we import SharedModule here? It includes things that we don't use like UnlinkGoogleAccountConfirmComponent and UnlinkGoogleAccountPasswordComponent. Instead, can we import only the classes that we need?

File: src/app/teacher/teacher.module.ts [hunk 3]
```diff
@@ -43,12 +43,18 @@ import { MatListModule } from '@angular/material/list';
...
+import { SelectTagsComponent } from './select-tags/select-tags.component';
+import { MatChipsModule } from '@angular/material/chips';
+import { SelectedTagsListComponent } from './selected-tags-list/selected-tags-list.component';
 
 const materialModules = [
   MatAutocompleteModule,
   MatButtonModule,
   MatCardModule,
   MatCheckboxModule,
+  MatChipsModule,
```
Comment (hirokiterashima) [inline src/app/teacher/teacher.module.ts]: Is this import necessary?

File: src/app/modules/library/select-all-items-checkbox/select-all-items-checkbox.component.spec.ts [hunk 4]
```diff
@@ -9,8 +9,7 @@ describe('SelectAllItemsCheckboxComponent', () => {
 
   beforeEach(() => {
     TestBed.configureTestingModule({
-      declarations: [SelectAllItemsCheckboxComponent],
-      imports: [MatCheckboxModule, MatTooltipModule]
+      imports: [MatCheckboxModule, MatTooltipModule, SelectAllItemsCheckboxComponent]
```
Comment (hirokiterashima) [inline src/app/modules/library/select-all-items-checkbox/select-all-items-checkbox.component.spec.ts]: Are MatCheckboxModule and MatTooltipModule needed?

File: src/app/teacher/color-chooser/color-chooser.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import { CommonModule } from '@angular/common';
+import { MatChipsModule } from '@angular/material/chips';
+
+describe('ColorChooserComponent', () => {
+  let component: ColorChooserComponent;
+  let fixture: ComponentFixture<ColorChooserComponent>;
+
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      imports: [ColorChooserComponent, CommonModule, MatChipsModule]
```
Comment (hirokiterashima) [inline src/app/teacher/color-chooser/color-chooser.component.spec.ts]: Are CommonModule and MatChipsModule needed?

File: src/app/teacher/edit-tag/edit-tag.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  let component: EditTagComponent;
+  let fixture: ComponentFixture<EditTagComponent>;
+
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      imports: [
+        BrowserAnimationsModule,
+        EditTagComponent,
+        HttpClientTestingModule,
+        MatSnackBarModule
```
Comment (hirokiterashima) [inline src/app/teacher/edit-tag/edit-tag.component.spec.ts]: This this import needed?

File: src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.spec.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+describe('ManageTagsDialogComponent', () => {
+  let component: ManageTagsDialogComponent;
+  let fixture: ComponentFixture<ManageTagsDialogComponent>;
+
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      imports: [
+        HttpClientTestingModule,
+        ManageTagsDialogComponent,
+        MatSnackBarModule,
```
Comment (hirokiterashima) [inline src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.spec.ts]: Is this import needed?

File: src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+  protected idToEditing: { [id: string]: boolean } = {};
+  protected inputChanged: Subject<any> = new Subject<any>();
+  protected showCreateTag: boolean;
+  private subscriptions: Subscription = new Subscription();
+  protected tags: Tag[] = [];
+
+  constructor(
+    private snackBar: MatSnackBar,
+    private teacherService: TeacherService,
+    private projectTagService: ProjectTagService
```
Comment (hirokiterashima) [inline src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.ts]: alphabetize?

**Review by hirokiterashima — CHANGES_REQUESTED** [review]
The functionality mostly works. 

It looks like the number of tags in the message when you try to delete a tag is off. In the screenshot below, you can see that I have "tag3" on a unit, but the cofirmation message says "0" I think it may not be counting the tags that are on the units (as opposed to run units).

Screenshot 2024-05-24 at 3 30 00 PM


I also added some code cleaning suggestions inline. 

Lastly, maybe we should also apply the code cleaning conventions for at least the new standalone components, like alphabetizing (imports -> selector -> standalone...), rename "styleUrls" -> "styleUrl", self-closing tags, and such?

**Comment by geoffreykwan** [comment]
I've fixed the bug and made all the changes.

Make sure to get the latest from the WISE-API branch issue-259-implement-unit-tagging for the fix to work.

**Review by hirokiterashima — CHANGES_REQUESTED** [review]
When I delete a tag in my unit library, the tag still appears on the unit. I have to refresh the page to see the change. Can they be immediately removed without having to refresh?

The tag choices in the "Apply Tags" filter dialog doesn't seem to update immediately after you add or delete tags. It only seems to update when you edit the tag name or color.

**Comment by geoffreykwan** [comment]
I've fixed the tags not updating bugs.

**Comment by hirokiterashima** [comment]
I verified the new commit fixes the tags not updating bugs.

The unit tests are failing now. I see a lot of ```NullInjectorError: No provider for ProjectTagService!``` errors.

For all the new html files that were introduced in this PR, (manage-tags-dialog.component.html, color-chooser.component.html, etc) should we apply the same improvements that we've been making in our standalone component work? I think doing this will make the template easier to read. Mainly:
- use new control flow syntax (```@if, @else if, @for,...```)
- use self-closing tags

**Comment by geoffreykwan** [comment]
Ok, I've fixed the tests, used the new flow control, and used self closing tags.

File: src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.html [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    @if (!showCreateTag) {
+      <div>
+        <button mat-flat-button color="primary" (click)="showCreateTag = !showCreateTag" i18n>
+          + New Tag
+        </button>
+      </div>
+    }
+    @if (showCreateTag) {
+      <edit-tag (closeEvent)="showCreateTag = false" />
+    }
```
Comment (hirokiterashima) [inline src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.html]: Switch test order and use ```@else``` for better readability?
```suggestion
    @if (showCreateTag) {
      <edit-tag (closeEvent)="showCreateTag = false" />
    } @else {
      <div>
        <button mat-flat-button color="primary" (click)="showCreateTag = !showCreateTag" i18n>
          + New Tag
        </button>
      </div>
    }
```

File: src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.html [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+      @for (tag of tags; track tag.id; let tagIndex = $index; let last = $last) {
+        <div>
+          <mat-list-item>
+            <div fxLayout="row" fxLayoutAlign="start center" fxLayoutGap="8px">
+              @if (idToEditing[tag.id]) {
+                <div class="edit-tag">
+                  <edit-tag [tag]="tag" (closeEvent)="cancelEditing(tag)" />
+                </div>
+              }
+              @if (!idToEditing[tag.id]) {
```
Comment (hirokiterashima) [inline src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.html]: Use ```@else``` here?
```suggestion
              } @else {
```

**Review by hirokiterashima — COMMENTED** [review]
New changes look good. I added some more code suggestions inline to improve control flow readability.

Also, I noticed that in the "Apply Tags" select dropdown, clicking on the choice span (not the checkbox, not the tag text, but the space next to those on the same row) does not add/delete the tag. The cursor is a pointer, so I expected it to add/delete. We should handle clicking on the entire row?

**Comment by geoffreykwan** [comment]
I've changed the flow control to use @if/@else and fixed the apply tags option bug.

**Review by hirokiterashima — APPROVED** [review]
LGTM

**Review by breity — APPROVED** [review]
Looks good. 👍

**Pull request merged**